onboarding_project/squad_pantry_app/admin.py

```diff
@@ -22,11 +31,150 @@ class OrderDishInline(admin.StackedInline):
+        except Order.DoesNotExist:
+            self.message_user(request, 'Mentioned Order ID does not exists', messages.ERROR)
+            return HttpResponseRedirect(request.META.get('HTTP_REFERER', '/admin/squad_pantry_app/order'))
+        else:
+            cancelled = order.cancel_order(request.user.id)
+            if cancelled == order.CANCEL_SUCCESS:
+                self.message_user(request, 'Order cancelled', messages.SUCCESS)
+            elif cancelled == order.WRONG_USER:
+                self.message_user(request, 'You do not have permission to cancel this order', messages.ERROR)
+            elif cancelled == order.PROCESSING_ORDER:
+                self.message_user(request, 'Could not cancel, The order is already being processed', messages.ERROR)
+            elif cancelled == order.ORDER_CLOSED_ERROR:
+                self.message_user(request, 'The order is already closed', messages.ERROR)
+
+            return HttpResponseRedirect(request.META.get('HTTP_REFERER', '/'))
+
+    def save_model(self, request, obj, form, change):
+        if not change:
+            obj.placed_by = request.user
+        return super(OrderAdmin, self).save_model(request, obj, form, change)
+
+    def get_readonly_fields(self, request, obj=None):
+        user = request.user
+        if obj is None:
+            if not user.is_kitchen_staff or (user.is_kitchen_staff and obj.status in obj.CLOSED_ORDERS
+                                             and obj.closed_at is not None):
+                return self.readonly_fields + ('status', )
+            return self.readonly_fields
+        else:
+            if not user.is_kitchen_staff or (user.is_kitchen_staff and obj.status in obj.CLOSED_ORDERS
+                                             and obj.closed_at is not None):
+                return self.readonly_fields + ('status', 'scheduled_time')
+            return self.readonly_fields + ('scheduled_time', )
+
+    def get_queryset(self, request):
+        qs = super(OrderAdmin, self).get_queryset(request)
+        if not request.user.is_kitchen_staff:
+            return qs.filter(placed_by=request.user)
+        return qs.filter()
+
+
+class UserCreationForm(forms.ModelForm):
+    password1 = forms.CharField(label='Password', widget=forms.PasswordInput())
+    password2 = forms.CharField(label='Password confirmation', widget=forms.PasswordInput())
+
+    class Meta:
+        model = SquadUser
+        fields = ('email', 'is_kitchen_staff')
+
+    def clean_password2(self):
+        password1 = self.cleaned_data.get("password1")
+        password2 = self.cleaned_data.get("password2")
+        if password1 and password2 and password1 != password2:
+            raise forms.ValidationError("Passwords don't match")
+        return password2
+
+    def save(self, commit=True):
+        user = super(UserCreationForm, self).save(commit=False)
+        user.set_password(self.cleaned_data["password1"])
+        if commit:
+            user.save()
```

Comment:
Now this entails 2 trips to the db. One to save the user initially, and then to set the password of the user and save again. Can we figure out a way to do this with just one hit on the db?

Reply:
You need a `user` object for the first save and you need save only if there's `commit`.